libebpfdiscoveryskel/src/LibSSLProbes.h

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2023 Dynatrace LLC
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#pragma once
+
+#include "DebugPrint.h"
+#include "GlobalData.h"
+#include "Handlers.h"
+#include "LibSSLTypes.h"
+#include "Log.h"
+#include "SysTypes.h"
+#include "TrackedSession.h"
+
+#include "ebpfdiscoveryshared/Constants.h"
+#include "vmlinux.h"
+
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_helpers.h>
+
+/*
+ * Maps for storing probed function arguments to pass them from uprobes to uretprobes.
+ */
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, __u64); // pid_tgid
+	__type(value, struct LibSSLReadArgs);
+	__uint(max_entries, DISCOVERY_MAX_SESSIONS);
+} runningLibSSLReadArgsMap SEC(".maps");
+
+/*
+ * Probe handlers
+ */
+
+int handleSSLReadEntry(struct pt_regs* ctx, void* ssl, char* buf) {
+	struct DiscoveryGlobalState* globalStatePtr = getGlobalState();
+	if (globalStatePtr == NULL || globalStatePtr->isCollectingDisabled) {
+		return 0;
+	}
+
+	if (ssl == NULL || buf == NULL) {
+		return 0;
+	}
+
+	struct LibSSLReadArgs sslReadArgs = {};
+	sslReadArgs.ssl = ssl;
+	sslReadArgs.buf = buf;
+
+	__u64 pidTgid = bpf_get_current_pid_tgid();
+	bpf_map_update_elem(&runningLibSSLReadArgsMap, &pidTgid, &sslReadArgs, BPF_ANY);
+	DEBUG_PRINTLN("ssl read entry, pid: `%d`", pidTgidToPid(pidTgid));
+
+	return 0;
+}
+
+int handleSSLReadExit(struct pt_regs* ctx, int bytesCount) {
+	struct DiscoveryGlobalState* globalStatePtr = getGlobalState();
+	if (globalStatePtr == NULL || globalStatePtr->isCollectingDisabled) {
+		return 0;
+	}
+
+	struct DiscoveryAllSessionState* allSessionStatePtr = getAllSessionState();
+	if (allSessionStatePtr == NULL) {
+		return 0;
+	};
+
+	__u64 pidTgid = bpf_get_current_pid_tgid();
+
+	struct LibSSLReadArgs* sslReadArgsPtr = (struct LibSSLReadArgs*)bpf_map_lookup_elem(&runningLibSSLReadArgsMap, &pidTgid);
+	if (sslReadArgsPtr == NULL) {
+		return 0;
+	}
+
+	if (sslReadArgsPtr->ssl == NULL || sslReadArgsPtr->buf == NULL) {
+		return 0;
+	}
+
+	if (bytesCount <= 0) {
```

Comment:
Also SSL_pending() can be used. In this case we won't hit this condition. Perhaps we should add uprobe for that?

Reply:
I don't see how SSL_pending() would influence the return value of SSL_read(). If it's true, please back your statement up with reference or documentation.

Reply:
> The read operation was not successful, because either the connection was closed, an error occurred or action must be taken by the calling process. Call [SSL_get_error(3)](https://www.openssl.org/docs/man1.0.2/man3/SSL_get_error.html) with the return value ret to find out the reason.

~ https://www.openssl.org/docs/man1.0.2/man3/SSL_read.html

If a server implementation checks with SSL_pending() that there is no more data to read we will never have SSL_read() call which returns value <= 0. Example:
```
// Loop to continuously read data until no more pending data
while (1) {
    int bytes_pending = SSL_pending(ssl);
    if (bytes_pending > 0) {
        // There is pending data to be read
        int bytes_read = SSL_read(ssl, buffer, sizeof(buffer));
        if (bytes_read > 0) {
            // Successfully read encrypted data
            // Process the decrypted data stored in 'buffer'
        } else if (bytes_read == 0) {
            // Connection was closed by peer
        } else {
            // Error occurred during SSL_read
            // Check SSL_get_error() for error details
            int error = SSL_get_error(ssl, bytes_read);
            // Handle the error appropriately
        }
    } else if (bytes_pending == 0) {
        // No pending data, exit the loop
        break;
    } else {
        // Error occurred while checking pending data
        // Handle the error appropriately
    }
}
```

Reply:
Sending the no-more-data event is done on best-effort basis, but it can be implemented for this case too. Thank you for the example.